### PR TITLE
[DTensor][autograd] Preserve DTensor type for the gradients of unused output

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #173295
* #173865

This is a tentative fix to https://github.com/pytorch/pytorch/issues/173123 and the issue encountered in https://github.com/pytorch/pytorch/pull/169393/, following the fix to NJT, https://github.com/pytorch/pytorch/pull/136875. When a DTensor is detected, zeros_like() will be used to create the zero gradients to preserve the tensor subclass type.

When an op has multiple outputs but not all are used (e.g., LSE for FlexAttention), autograd will materialize the gradient of the unused output as a zero tensor. But autograd doesn't respect the tensor   subclass. In our case, the q, k, v to FlexAttention are all DTensors, hence the outputs of FlexAttention, including LSE, are all DTensors. But LSE is not always used (depending on the model), which results in errors in both eager mode and compiler mode.

In eager mode, DTensor complains users mix torch.Tensor with DTensor in some computation. In compiler mode, compiler raises an exception due to an expected metadata mismatch.

Semantically this fix is correct since we should not upcast a tensor back to a plain tensor. To limit the blast area, we specifically detect DTensor rather than all tensor subclasses, to avoid affecting other subclasses that may not need this behavior.

**Conclusion**
Based on the discussion with @drisspg, @bdhirsh, @soulitzer, @wconstab, we can land this PR as we limit the PR to DTensor only. But we also need to add a warning to indicate the potential performance issue and point out the solution to remove the zeros_like(). The warning will also print out the backward name so that users can debug.

cc @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx @weifengpy

### CODE REVIEW

**Review by aaron-seq — CHANGES_REQUESTED** [review]
The implementation correctly identifies the need to preserve tensor subclass types for unused gradients, but the detection mechanism is too broad, causing 19 test failures.

**Root cause:**
Using `is_python_dispatch()` (line 783) catches all Python dispatch tensors, including functorch's internal tensors (FakeTensor, AOTAutograd wrappers). When `zeros_like()` is applied to these tensors instead of creating plain zero tensors, it alters gradient materialization behavior and breaks functorch's input mutation tracking.

**Failing tests:**
- 18 failures in `test/functorch/test_aotdispatch.py` (input mutation tests)
- 1 failure in `test/test_autograd.py::test_undefined_grad_with_tensor_subclass` (the new test added in this PR)

**Recommendation:**
Replace the broad `is_python_dispatch()` check with more specific detection for DTensor and other tensor subclasses that require `zeros_like()` preservation. Consider checking for specific subclass types or adding a trait that indicates whether a tensor subclass needs preserved zero gradient behavior, rather than applying this to all Python dispatch tensors.

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -769,14 +769,20 @@ static void _wrap_outputs(
...
         bool is_differentiable =
             (non_differentiable.count(wrapped_output->unsafeGetTensorImpl()) ==
                  0 &&
              isDifferentiableType(wrapped_output->scalar_type()));
-        bool use_zeros_like =
-            is_differentiable && num_outputs > 1 && wrapped_output->is_nested();
+        bool is_tensor_subclass =
+            wrapped_output->unsafeGetTensorImpl()->is_python_dispatch();
+        bool use_zeros_like = is_differentiable && num_outputs > 1 &&
+            (wrapped_output->is_nested() || is_tensor_subclass);
```
Comment (wconstab) [inline torch/csrc/autograd/python_function.cpp]: What does is_nested mean? (Just curious)

Comment (fegin) [reply torch/csrc/autograd/python_function.cpp]: The tensor is NJT.

File: torch/csrc/autograd/python_function.cpp [hunk 2]
```diff
@@ -769,14 +770,34 @@ static void _wrap_outputs(
...
              isDifferentiableType(wrapped_output->scalar_type()));
-        bool use_zeros_like =
-            is_differentiable && num_outputs > 1 && wrapped_output->is_nested();
+        bool output_is_dtensor = is_dtensor(obj);
+        bool use_zeros_like = is_differentiable && num_outputs > 1 &&
+            (wrapped_output->is_nested() || output_is_dtensor);
+        if (use_zeros_like && output_is_dtensor) {
+          TORCH_WARN_ONCE(
+              "Autograd is calling zeros_like() on a DTensor to materialize "
+              "the gradient for an unused output in ",
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp]:
```suggestion
              "the gradient for an unused output in the autograd.Function named: ",
```

File: torch/csrc/autograd/python_function.cpp [hunk 3]
```diff
@@ -769,14 +770,34 @@ static void _wrap_outputs(
...
+              cdata->name(),
+              ". This preserves the DTensor type but may have performance "
+              "implications. To avoid this: (1) call "
+              "ctx.set_materialize_grads(False) in forward() (only works when "
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp]: Suggesting a different wording here:

To avoid this:
1) If you are in eager (e.g. the name of the autograd.Function is not CompiledFunction) it means that one of the outputs of your autograd.Function is unused. You can call ctx.set_materialize_grads(False) in the forward() and update the backward() to handle potential None gradients.
2) If you are in compile, it means that one of the outputs to the compiled region is unused. You should detach() any outputs that are unused and explicitly or implicitly (i.e., saved as a global) returned from the compiled region.

**Review by soulitzer — APPROVED** [review]
Thanks, add some small comments.

File: torch/csrc/autograd/python_variable.cpp [hunk 4]
```diff
@@ -850,6 +850,21 @@ py::handle get_dtensor_class() {
   return get_dtensor_class_impl();
 }
 
+bool is_dtensor(PyObject* obj) {
+#ifdef USE_DISTRIBUTED
```
Comment (wconstab) [inline torch/csrc/autograd/python_variable.cpp:855]: wonder if we should delete this? i think the rest of DTensor code still exists even without this, but i'm not sure exactly what our policy is?

File: torch/csrc/autograd/python_function.cpp [hunk 5]
```diff
@@ -786,17 +786,33 @@ static void _wrap_outputs(
...
-              "the autograd.Function is not compiled with torch.compile), "
-              "(2) detach unused outputs before returning from forward() "
-              "(required if torch.compile() may be used), and (3) handle None "
-              "gradients in backward() instead of relying on zero tensors.");
+              "the gradient for an unused output in autograd.Function '",
+              node_name,
+              "'. This preserves the DTensor type but may have performance "
+              "implications. To avoid this:\n"
+              "(1) In eager mode (i.e., the function name is not "
+              "'CompiledFunction' or 'ApplyTemplate'): One of the "
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp]: I think ApplyTemplate DOES fall into the category where you have a problem that exists in eager. And then we can instruct the user to run it again w/o compile to get the actual name.

File: torch/csrc/autograd/python_variable.cpp (hunk 4 above)
Comment (fegin) [reply torch/csrc/autograd/python_variable.cpp:855]: I prefer to keeping this as this API is also used by some other logic which I don't have enough backgroud to make the call.

Comment (wconstab) [reply torch/csrc/autograd/python_variable.cpp:855]: i think you should check what happens if you try to import the class DTensor in a torch build without USE_DISTRIBUTED.  if DTensor still imports, then we shouldn't use the macro here. Otherwise we should? (it should be consistent)

File: test/distributed/tensor/test_dtensor.py [hunk 6]
```diff
@@ -788,6 +788,74 @@ def test_shard_tensor_2d(self):
...
+                        # ctx.set_materialize_grads(False)
+                        return (x * 2), (y * 3)
+
+                    @staticmethod
+                    def backward(ctx, grad_out1, grad_out2):
+                        assert isinstance(grad_out2, DTensor)
+                        if not use_compile:
+                            MultiOutputFunc.gout1 = grad_out1
+                            MultiOutputFunc.gout2 = grad_out2
+                        return grad_out1 * 2, grad_out2
```
Comment (drisspg) [inline test/distributed/tensor/test_dtensor.py]: nit fix grad2 calc

File: torch/csrc/autograd/python_function.cpp [hunk 7]
```diff
@@ -769,14 +770,54 @@ static void _wrap_outputs(
...
+        // Note: We specifically check for DTensor rather than all tensor
+        // subclasses to avoid affecting other subclasses that may not need
+        // this behavior.
         bool is_differentiable =
             (non_differentiable.count(wrapped_output->unsafeGetTensorImpl()) ==
                  0 &&
              isDifferentiableType(wrapped_output->scalar_type()));
-        bool use_zeros_like =
-            is_differentiable && num_outputs > 1 && wrapped_output->is_nested();
+        bool output_is_dtensor = is_dtensor(obj);
```
Comment (drisspg) [inline torch/csrc/autograd/python_function.cpp:795]: @claude what do you think it feels bad that we have specilaized this path to two subclasses. Do you think we should define some sort of more generic interface here that subclasses can opt into?

File: test/distributed/tensor/test_dtensor.py [hunk 8]
```diff
@@ -788,6 +788,203 @@ def test_shard_tensor_2d(self):
...
+                        gout2.placements,
+                        out2.placements,
+                        "grad_out2 should have the same placements as out2",
+                    )
+
+    @with_comms
+    def test_multi_layer_memory_usage_with_zeros_like(self):
+        """
+        Test memory usage when multiple layers each use MultiOutputFunc with
+        an unused DTensor output. Selective activation checkpointing (SAC) is
```
Comment (soulitzer) [inline test/distributed/tensor/test_dtensor.py:925]: The SAC interaction is pretty orthogonal imo, to make the intent of the test clearer, just test w/o SAC, and check that the output is not saved?

File: torch/csrc/autograd/python_function.cpp [hunk 9]
```diff
@@ -771,15 +780,36 @@ static void _wrap_outputs(
...
         bool is_differentiable =
             (non_differentiable.count(wrapped_output->unsafeGetTensorImpl()) ==
                  0 &&
              isDifferentiableType(wrapped_output->scalar_type()));
-        bool use_zeros_like =
-            is_differentiable && num_outputs > 1 && wrapped_output->is_nested();
-        self->output_info.emplace_back(wrapped_output.value(), use_zeros_like);
+        bool output_is_dtensor = is_dtensor(obj);
+        bool output_is_nested = wrapped_output->is_nested();
+        bool use_zeros_like = is_differentiable && num_outputs > 1 &&
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp:797]: Do we want to combine the next PR (num_outputs = 1 case) into this one?

**Review by soulitzer — APPROVED** [review]
Thanks! Added minor comments

File: torch/csrc/autograd/variable_info.cpp [hunk 10]
```diff
@@ -23,10 +23,21 @@ VariableInfo::VariableInfo(const Variable& var, bool use_zeros_like)
 
 VariableInfo::VariableInfo() : requires_grad(false), is_empty(true) {}
 
+void VariableInfo::save_template_for_zeros(const Variable& var) {
+  // Create a scalar (0-dim) tensor that preserves the tensor subclass type.
+  // For DTensor, this preserves the mesh and placements while using minimal
+  // memory (just a single element instead of the full tensor).
+  // We use new_zeros({}) to create a scalar with the same type/subclass.
+  template_var = var.new_zeros({}).detach();
```
Comment (soulitzer) [inline torch/csrc/autograd/variable_info.cpp:31]: Detach before new_zeros
```suggestion
  template_var = var.detach().new_zeros({});
```
